### ISSUE TITLE
`stat_binhex` hexagons appear extremely stretched if all data points fall in a single bin on y dimension

### ISSUE THREAD

**Issue by ail24** [tidyverse/ggplot2#4463]
If all data points are in a single bin on the y axis (e.g. if they are all equal to 0) the hexagons appear extremely stretched vertically. 

Minimal reprex:
```r
data <- tibble(x = rnorm(1000, 20, 20), y = rep(0,1000))

data %>%
  ggplot(aes(x = x, y = y))+
  stat_binhex(bins = 100)+
  scale_x_continuous(limits = c(-100,100))+
  scale_y_continuous(limits = c(-100,100))
```

![plot2](https://user-images.githubusercontent.com/57092091/117078143-99341300-ad39-11eb-854c-3976da0a33de.png)

**Comment by smouksassi** [comment]
for stat_binhex the provided limits will be used for computing the ybns

ybnds <- hex_bounds(y, binwidth[2])

what output do you expect? you might want stat count where you are simply counting how many y at each unique x position

**Comment by clauswilke** [comment]
Something is definitely wrong when the variation in the data is less than the bin width in that direction. Even if `binwidth` is provided explicitly the plot fails.

``` r
library(tidyverse)

set.seed(1234)
data <- tibble(
  x = rnorm(1000, 20, 20),
  y1 = rnorm(1000, 0, 2),
  y2 = rnorm(1000, 0, .2)
)

data %>%
  ggplot(aes(x = x, y = y1)) +
  stat_binhex() +
  scale_x_continuous(limits = c(-100, 100)) +
  scale_y_continuous(limits = c(-100, 100))
```

![](https://i.imgur.com/XjX9eVR.png)

``` r
data %>%
  ggplot(aes(x = x, y = y2)) +
  stat_binhex() +
  scale_x_continuous(limits = c(-100, 100)) +
  scale_y_continuous(limits = c(-100, 100))
```

![](https://i.imgur.com/OEFBdyx.png)

``` r
data %>%
  ggplot(aes(x = x, y = y1)) +
  stat_binhex(binwidth = c(5, 5)) +
  scale_x_continuous(limits = c(-100, 100)) +
  scale_y_continuous(limits = c(-100, 100))
```

![](https://i.imgur.com/4ttIUkK.png)

``` r
data %>%
  ggplot(aes(x = x, y = y2)) +
  stat_binhex(binwidth = c(5, 5)) +
  scale_x_continuous(limits = c(-100, 100)) +
  scale_y_continuous(limits = c(-100, 100))
```

![](https://i.imgur.com/m6YVfT7.png)

``` r
# without limits

data %>%
  ggplot(aes(x = x, y = y2)) +
  stat_binhex()
```

![](https://i.imgur.com/o1jHavO.png)

``` r
data %>%
  ggplot(aes(x = x, y = y2)) +
  stat_binhex(binwidth = c(5, 5))
```

![](https://i.imgur.com/ead0OCB.png)

<sup>Created on 2021-05-05 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by clauswilke** [comment]
My previous post wasn't entirely correct. Manually setting the bin height does work when limits aren't set. I just had picked a poor value in my previous example.

``` r
library(tidyverse)

set.seed(1234)
data <- tibble(
  x = rnorm(1000, 20, 20),
  y1 = rnorm(1000, 0, 2),
  y2 = rnorm(1000, 0, .2)
)

data %>%
  ggplot(aes(x = x, y = y2)) +
  stat_binhex(binwidth = c(5, 0.08))
```

![](https://i.imgur.com/GJv5qgm.png)

<sup>Created on 2021-05-05 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by thomasp85** [comment]
Fixed by #4671
